### PR TITLE
feat: Linux & Python Runtime Support

### PR DESCRIPTION
**Pull request by tbarlow12**
## What did you implement:

Runtime support for Linux and Python

## How did you implement it:

- Added `kind` to App Service plan and Function app ARM templates
- Added `reserved` and `linuxFxVersion` to Function App template
- Moved `spawn` to Utils class
- Added `CoreToolsService` to handle any CLI invocation of `func`
- Using `CoreToolsService` to package & publish Linux apps
- Only cleaning up built files when not publishing to Linux (see note below)
- Forces Python to run on Linux OS
- Use `/` instead of OS-specific path separator for `function.json` file (works for both windows and linux)
- Cleaning up generated python files in package service
- Fixed typo in name `uploadZippedArfifactToFunctionApp`

*Important*: Because core tools does not currently support publishing an existing package, any Linux applications will not clean built files. Also, because `func azure` (used in publishing) invokes the Azure CLI, I have not been able to get it to run properly as a child process. This will currently give the user the exact command to both publish their function app and then clean up the generated files.

## How can we verify it:

Specify the `linux` os in your `serverless.yml` like:

```yaml
provider:
    os: linux
```

and run `sls deploy`.

## Todos:

_**Note: Run `npm run test:ci` to run all validation checks on proposed changes**_

- [x] Ensure there are no lint errors.  
       **Validate via `npm run lint`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
- [x] Write tests and confirm existing functionality is not broken.  
       **Validate via `npm test`**
- [x] Write documentation
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES  
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
   "typings": "./lib/index.d.ts",
   "author": "Azure Functions",
   "scripts": {
-    "start": "npm run test -- --watch",
+    "start": "jest --watch",
```
Comment (wbreza) [inline package.json]: why the change to trigger jest directly vs. the NPM command.  It doesn't seem like a big change but now `pretest` etc are also not run which did some linting, etc.

File: src/armTemplates/resources/functionApp.test.ts [hunk 2]
```diff
@@ -1,51 +1,126 @@
...
 
+  const defaultConfig: ServerlessAzureConfig = {
+    provider: {
+      name: "azure",
+      prefix,
+      region,
+      stage,
+      resourceGroup: resourceGroupName,
+      functionRuntime: {
+        language: SupportedRuntimeLanguage.NODE,
```
Comment (wbreza) [inline src/armTemplates/resources/functionApp.test.ts]: Does this interfere with setting the os / runtime directly on the provider?  What are we recommending to users?  Why do we have 2 ways?

File: src/armTemplates/resources/functionApp.test.ts [hunk 3]
```diff
@@ -1,51 +1,126 @@
...
+    });
+  
+    it("gets correct linux parameters - node 8", () => {
+      const config: ServerlessAzureConfig = {
+        ...defaultConfig,
+        provider: {
+          ...defaultConfig.provider,
+          os: FunctionAppOS.LINUX,
+          functionRuntime: {
+            language: SupportedRuntimeLanguage.NODE,
```
Comment (wbreza) [inline src/armTemplates/resources/functionApp.test.ts]: Same as above - how do I know what param to set in the yaml?

File: src/armTemplates/resources/functionApp.ts [hunk 4]
```diff
@@ -39,6 +43,18 @@ export class FunctionAppResource implements ArmResourceTemplateGenerator {
         defaultValue: "",
         type: ArmParamType.String
       },
+      functionAppKind: {
+        defaultValue: "functionapp",
+        type: ArmParamType.String,
+      },
+      functionAppReserved: {
+        defaultValue: false,
```
Comment (wbreza) [inline src/armTemplates/resources/functionApp.ts]: We should add some inline comments in the ARM template resources that describe the purpose of each of these properties.

File: src/config.ts [hunk 5]
```diff
@@ -40,7 +46,28 @@ export const configConstants = {
...
+  dockerImages: {
+    node: {
+      "8": "DOCKER|microsoft/azure-functions-node8:2.0",
+      "10": "DOCKER|microsoft/azure-functions/node:2.0"
+    },
+    python: {
+      "3.6": "DOCKER|microsoft/azure-functions/python:2.0"
+    }
   },
+  defaultFuncIgnore: [
```
Comment (wbreza) [inline src/config.ts]: Does this play nice with Node apps on linux?  If I don't use webpack I would need to ensure my `package.json` and `node_modules` are included in my deployment.

File: src/plugins/azureBasePlugin.ts [hunk 6]
```diff
@@ -10,6 +10,10 @@ export abstract class AzureBasePlugin<TOptions=Serverless.Options> {
   protected config: ServerlessAzureConfig;
   protected commands: ServerlessCommandMap;
   protected processedCommands: ServerlessCliCommand[];
+  /**
+   * Specifies if targeted runtime is linux
+   */
+  protected isLinuxTarget: boolean;
```
Comment (wbreza) [inline src/plugins/azureBasePlugin.ts]: Doesn't feel like this property should exist on the base plugin since its a concern of deployment

File: src/plugins/login/utils/simpleFileTokenCache.test.ts [hunk 7]
```diff
@@ -1,9 +1,12 @@
-import fs from "fs";
 import path from "path";
 import os from "os";
 import mockFs from "mock-fs";
 import { MockFactory } from "../../../test/mockFactory";
 import { SimpleFileTokenCache } from "./simpleFileTokenCache";
+let fs;
+jest.isolateModules(() => {
```
Comment (wbreza) [inline src/plugins/login/utils/simpleFileTokenCache.test.ts]: What does the `isolateModules()` do?

Comment (wbreza) [reply src/plugins/login/utils/simpleFileTokenCache.test.ts]: Is it needed since you figured out the mock/restore stuff?

File: src/services/baseService.test.ts [hunk 8]
```diff
@@ -1,10 +1,14 @@
-import fs from "fs";
 import mockFs from "mock-fs";
 import Serverless from "serverless";
 import { ServerlessAzureOptions } from "../models/serverless";
 import { MockFactory } from "../test/mockFactory";
 import { BaseService } from "./baseService";
 
+let fs;
```
Comment (wbreza) [inline src/services/baseService.test.ts]: Confirm if this complexity is needed.

File: src/services/baseService.ts [hunk 9]
```diff
@@ -38,6 +41,9 @@ export abstract class BaseService {
     this.resourceGroup = this.configService.getResourceGroupName();
     this.deploymentName = this.configService.getDeploymentName();
     this.artifactName = this.configService.getArtifactName(this.deploymentName);
+    this.runtime = this.configService.getRuntime();
+    this.pythonTarget = this.runtime.language === SupportedRuntimeLanguage.PYTHON;
```
Comment (wbreza) [inline src/services/baseService.ts]: Similar to above - doesn't feel like python/linux related variables should exist on the base service.  Also, since they are boolean variables i would rename to something like `isPythonTarget` && `isLinuxTarget`

File: src/services/functionAppService.ts [hunk 10]
```diff
@@ -157,51 +158,19 @@ export class FunctionAppService extends BaseService {
...
-        functionApp,
-        configConstants.runFromPackageSetting,
-        sasUrl
-      );
-
-      await this.syncTriggers(functionApp, response.properties);
+    // Uploaded to blob storage as artifact for future reference
+    await this.uploadZippedArtifactToBlobStorage(functionZipFile);
+
+    if (this.linuxTarget) {
```
Comment (wbreza) [inline src/services/functionAppService.ts]: This is better - but I think this block should be encapsulated in a separate service so that the `functionAppService` doesn't need to care about the concerns of the OS.

File: src/services/functionAppService.ts [hunk 11]
```diff
@@ -223,7 +192,7 @@ export class FunctionAppService extends BaseService {
     return await this.get();
   }
 
-  public async uploadZippedArfifactToFunctionApp(functionApp: Site, functionZipFile: string) {
```
Comment (wbreza) [inline src/services/functionAppService.ts]: Arfifact 🤣

File: src/services/packageService.test.ts [hunk 12]
```diff
@@ -241,7 +247,97 @@ describe("Package Service", () => {
...
     copyFileSpy.mockRestore();
   });
+
+  describe("Python packages", () => {
+    let mySpawn;
+    let packageService: PackageService;
+    let functions: string[];
+
+    beforeEach(() => {
+
```
Comment (wbreza) [inline src/services/packageService.test.ts]: nit: empty lines

File: src/services/packageService.ts [hunk 13]
```diff
@@ -69,30 +88,80 @@ export class PackageService extends BaseService {
...
   public createBinding(functionName: string, functionMetadata: FunctionMetadata) {
-    const functionJSON = functionMetadata.params.functionsJson;
-    functionJSON.entryPoint = functionMetadata.entryPoint;
-    functionJSON.scriptFile = functionMetadata.handlerPath;
+    const functionJSON = this.getFunctionJson(functionName, functionMetadata);
+    const functionDirPath = this.makeFunctionDir(functionName);
+
+    fs.writeFileSync(path.join(functionDirPath, "function.json"), this.stringify(functionJSON));
+
+    if (this.runtime.language === SupportedRuntimeLanguage.PYTHON) {
```
Comment (wbreza) [inline src/services/packageService.ts]: What is happening here?  Writing an empty `__init__.py` file?  How does this work?  Would be good to have some comments here with intention of this change.

File: package.json (hunk 1 above)
Comment (tbarlow12) [reply package.json]: Was going to ask your thoughts on this. I thought it was kind of nice to have a command to just trigger tests without having to wait for linting to finish. It was taking kind of a while for me, but I'm open to leaving this as is

File: src/armTemplates/resources/functionApp.test.ts (hunk 2 above)
Comment (tbarlow12) [reply src/armTemplates/resources/functionApp.test.ts]: This is an object that I'm inferring from the `runtime` in the provider, and then setting in `functionRuntime`. I didn't want to change `runtime` directly, but I can if that would be better. I just didn't want to parse and instantiate this object every time

File: src/armTemplates/resources/functionApp.test.ts (hunk 3 above)
Comment (tbarlow12) [reply src/armTemplates/resources/functionApp.test.ts]: User would only set `runtime`

File: src/config.ts (hunk 5 above)
Comment (tbarlow12) [reply src/config.ts]: I'll update this.

File: src/plugins/azureBasePlugin.ts (hunk 6 above)
Comment (tbarlow12) [reply src/plugins/azureBasePlugin.ts]: I think this will go away when I address your other feedback on reducing some of the complexity

File: src/plugins/login/utils/simpleFileTokenCache.test.ts (hunk 7 above)
Comment (tbarlow12) [reply src/plugins/login/utils/simpleFileTokenCache.test.ts]: Nope. Good catch. I'll remove

File: src/services/packageService.ts (hunk 13 above)
Comment (tbarlow12) [reply src/services/packageService.ts]: I'll see if it works without it, but I've always had to create `__init__.py` in order for the directory to be considered a package. [Helpful article for more info](https://medium.com/@ramrajchandradevan/python-init-py-modular-imports-81b746e58aae#targetText=The%20__init__.py,on%20the%20module%20search%20path.)

File: src/armTemplates/resources/functionApp.test.ts (hunk 2 above)
Comment (tbarlow12) [reply src/armTemplates/resources/functionApp.test.ts]: Will add as private configuration within config service

File: src/armTemplates/resources/functionApp.test.ts (hunk 3 above)
Comment (tbarlow12) [reply src/armTemplates/resources/functionApp.test.ts]: Same as above

File: src/plugins/azureBasePlugin.ts (hunk 6 above)
Comment (tbarlow12) [reply src/plugins/azureBasePlugin.ts]: Add to config service

File: src/services/functionAppService.ts (hunk 10 above)
Comment (tbarlow12) [reply src/services/functionAppService.ts]: Will create a `functionAppPublishService`

File: src/armTemplates/resources/functionApp.test.ts (hunk 2 above)
Comment (tbarlow12) [reply src/armTemplates/resources/functionApp.test.ts]: I just realized that I can't do this ^. It needs to be inside the `ServerlessAzureConfig` because that's the only data that is passed to the ARM template resources like `functionApp`. Will add a comment on the property indicating that it's only for internal use

File: src/services/packageService.ts (hunk 13 above)
Comment (tbarlow12) [reply src/services/packageService.ts]: Removed this

File: src/armTemplates/resources/appServicePlan.ts [hunk 14]
```diff
@@ -71,10 +77,15 @@ export class AppServicePlanResource implements ArmResourceTemplateGenerator {
...
     };
 
+    const { os } = config.provider;
+
     const params: AppServicePlanParams = {
       appServicePlanName: {
         value: AppServicePlanResource.getResourceName(config),
       },
+      kind: {
+        value: (os === FunctionAppOS.LINUX) ? "Linux" : undefined,
```
Comment (pjlittle) [inline src/armTemplates/resources/appServicePlan.ts]: interesting - so undefined implies Windows?

Comment (tbarlow12) [reply src/armTemplates/resources/appServicePlan.ts]: The armService removes parameters with undefined parameters on deployment. So by keeping it undefined, it just won't be added

File: src/shared/utils.ts [hunk 15]
```diff
@@ -188,4 +200,54 @@ export class Utils {
...
+    }
+
+    const env = {
+      // Inherit environment from current process, most importantly, the PATH
+      ...process.env,
+      // Environment variables from serverless config are king
+      ...serverless.service.provider["environment"],
+    }
+    if (!options.silent) {
+      serverless.cli.log(`Spawning process '${command} ${commandArgs.join(" ")}'`);
```
Comment (mydiemho) [inline src/shared/utils.ts]: nit: we should refactor the log method to `log`, `info`, etc. and turn each on/off accordingly with the `silent` flag

File: src/armTemplates/resources/functionApp.ts [hunk 16]
```diff
@@ -1,14 +1,52 @@
...
+import { FunctionAppConfig, ServerlessAzureConfig, SupportedRuntimeLanguage, FunctionAppOS } from "../../models/serverless";
 import { AzureNamingService, AzureNamingServiceOptions } from "../../services/namingService";
+import configConstants from "../../config";
 
 interface FunctionAppParams extends DefaultArmParams {
+  /**
+   * Name of function app
+   */
   functionAppName: ArmParameter;
+  /**
```
Comment (mydiemho) [inline src/armTemplates/resources/functionApp.ts]: nit: new line before comments for readability, especially for text to speech reader

File: src/armTemplates/resources/functionApp.ts [hunk 17]
```diff
@@ -1,14 +1,52 @@
...
+   * Node version of function app
+   */
   functionAppNodeVersion: ArmParameter;
+  /**
```
Comment (mydiemho) [inline src/armTemplates/resources/functionApp.ts]: same for all the interface elements below

File: src/shared/utils.ts (hunk 15 above)
Comment (tbarlow12) [reply src/shared/utils.ts]: Yep, that's already in the [backlog](https://github.com/serverless/serverless-azure-functions/issues/374)

Comment (tbarlow12) [reply src/shared/utils.ts]: Although this `log` method is part of the serverless core, not something we can refactor.

File: src/armTemplates/resources/functionApp.ts (hunk 16 above)
Comment (tbarlow12) [reply src/armTemplates/resources/functionApp.ts]: Hmm, not sure. We haven't done that before, and we don't have new lines between the members of other interfaces. I don't know if a jsdoc comment should change that

**Comment by Vbubblery** [comment]
Hello,

May I ask when could you merge this PR?

File: src/config.ts [hunk 18]
```diff
@@ -40,8 +46,26 @@ export const configConstants = {
...
+  },
+  dockerImages: {
+    node: {
+      "8": "DOCKER|microsoft/azure-functions-node8:2.0",
+      "10": "DOCKER|microsoft/azure-functions/node:2.0"
+    },
+    python: {
+      "3.6": "DOCKER|microsoft/azure-functions/python:2.0"
+    }
   },
```
Comment (Hazhzeng) [inline src/config.ts]: In Linux, currently we accepts `PYTHON|3.6`, `PYTHON|3.7`, `PYTHON|3.8`, `NODE|8`, `NODE|10` in the **LinuxFxVersion**, and we will convert these settings to the proper images for you.

Placing a list of MCR docker images have been deprecated in LinuxFxVersion setting.

**Comment by tbarlow12** [comment]
Closing in favor of coming PR
